facilitator/src/work_queue.rs

```diff
@@ -0,0 +1,103 @@
+use anyhow::{anyhow, Result};
+use std::sync::{Arc, Mutex};
+
+/// A queue of jobs that may be shared across multiple consumer threads. The
+/// producer must have the entire list of jobs in hand when creating the
+/// WorkQueue; adding new jobs to an existing WorkQueue is not supported.
+#[derive(Debug, Clone)]
+pub(crate) struct WorkQueue<T, R> {
+    /// The list of jobs, wrapped in an std::sync::Mutex and std::sync::Arc for
+    /// thread safety. Most work queue implementations would use an
+    /// std::collections::VecDeque to allow new jobs to be pushed to the back of
+    /// the queue, but since we require clients to have all the jobs in hand
+    /// when they call WorkQueue::new(), we can just use a plain Vec and save an
+    /// allocation and copy.
+    jobs: Arc<Mutex<Vec<T>>>,
+    results: Arc<Mutex<Vec<R>>>,
+}
+
+impl<T, R> WorkQueue<T, R> {
+    /// Creates a new WorkQueue from the provided list of jobs
+    pub(crate) fn new(jobs: Vec<T>) -> Self {
+        WorkQueue {
+            jobs: Arc::new(Mutex::new(jobs)),
+            results: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+
+    /// Returns a job from the queue, or None if the queue is empty.
+    pub(crate) fn dequeue_job(&mut self) -> Option<T> {
+        // This and other methods on WorkQueue will panic if the Mutex is
+        // tainted, which means some other thread panicked while holding the
+        // lock.
+        // https://doc.rust-lang.org/std/sync/struct.Mutex.html#method.lock
+        self.jobs.lock().unwrap().pop()
+    }
+
+    /// Send job results to the queue. Clients need not call this 1:1 with
+    /// dequeue_job, if they can reasonably represent there results of multiple
+    /// jobs in one R.
+    pub(crate) fn send_results(&mut self, result: R) {
+        self.results.lock().unwrap().push(result)
+    }
+
+    /// Get the results from this work queue's completed jobs, consuming the
+    /// work queue. It is the caller's responsibility to ensure that no workers
+    /// are still dequeueing jobs.
```

Comment:
This can enforce its expectations a little more strongly: It can check that `self.jobs` is empty, upon pain of `Err()`.

Reply:
> Perhaps it's my Go leanings, but I would be somewhat tempted to use a [std::sync::mpsc::sync_channel](https://doc.rust-lang.org/stable/std/sync/mpsc/fn.sync_channel.html) to implement the results-collection side of `WorkQueue`.

That's a good idea! I had looked at `mpsc` as a replacement for all of `WorkQueue`, but realized that what I needed is `spmc`, which the stdlib doesn't provide. seanmonstar of Hyper fame has an [spmc crate](https://github.com/seanmonstar/spmc) but finally I decided against adding a dependency since `WorkQueue` is so simple. Anyway, I'll try using `std::sync::mpsc` for the results side and we'll see if it's any nicer.

Reply:
I gave this a try with `mpsc::SyncChannel` and found it to be. In my usage, I clone the `WorkQueue` into each worker thread so they can all send results. [`std::sync::mpsc::Receiver`](https://doc.rust-lang.org/stable/std/sync/mpsc/struct.Receiver.html) is not `Clone` because you're not meant to share it across threads. But what that means is that I can't have the send and receive ends of the queue abstracted behind a single object: the caller in `aggregation.rs` would have to hold on to both a `WorkQueue` (to be cloned off into worker threads) and a `std::sync::mpsc::Receiver` (to eventually get the results). Since using `mpsc::SyncChannel` doesn't make the implementation of `WorkQueue` any simpler, I'm going to stick with what I've got.